DolphinDynamicInputTexture/Data/InputSubRegionRect.cs

```diff
@@ -0,0 +1,168 @@
+                    LastMainRegionHeight = value.Height;
+                    _main_region = value;
+                }
+                MainRegion.PropertyChanged += OnMainRegionChanged;
+                OnPropertyChanged(nameof(MainRegion));
+            }
+        }
+
+        public IRectRegion MainRegion => _main_region;
+        private IRectRegion _main_region;
+
+        public new double X
+        {
+            get => MainRegion != null ? MainRegion.X + base.X : base.X;
+            set
+            {
+                if (MainRegion != null)
+                {
+                    base.X = value - MainRegion.X;
+
+                    if (RightX > MainRegion.RightX)
+                        base.X -= RightX - MainRegion.RightX;
+                }
+                else
+                {
+                    base.X = value;
+                }
+            }
+        }
+
+        public new double Y
+        {
+            get => MainRegion != null ? MainRegion.Y + base.Y : base.Y;
+            set
+            {
+                if (MainRegion != null)
+                {
+                    base.Y = value - MainRegion.Y;
+
+                    if (BottomY > MainRegion.BottomY)
+                        base.Y -= BottomY - MainRegion.BottomY;
+                }
+                else
+                {
+                    base.Y = value;
+                }
+            }
+        }
+
+        public new double Width
+        {
+            get => base.Width;
+            set
+            {
+                base.Width = value;
+                if (MainRegion != null)
+                {
+                    if (RightX > MainRegion.RightX)
+                        base.Width -= RightX - MainRegion.RightX;
+                }
+            }
+        }
+
+        public new double Height
+        {
+            get => base.Height;
+            set
+            {
+                base.Height = value;
+                if (MainRegion != null)
+                {
+                    if (BottomY > MainRegion.BottomY)
+                        base.Height -= BottomY - MainRegion.BottomY;
+                }
+            }
+        }
+
+        public new double BottomY => Y + Height;
+
+        public new double RightX => X + Width;
+
+        public new bool Equals([AllowNull] IRectRegion other)
+        {
+            return other != null && Equals(other.X, other.Y, other.Width, other.Height);
+        }
+
+        public new bool Equals(double x, double y, double width, double height)
+        {
+            return x == X & y == Y & width == Width & height == Height;
+        }
+
+        private double LastMainRegionWidth;
+        private double LastMainRegionHeight;
+
+        private void UpdateXScale()
+        {
+            if (LastMainRegionWidth != MainRegion.Width)
+            {
+                int i = InputRegionRect.DecimalPlaces;
+                InputRegionRect.DecimalPlaces = 3;
```

Comment:
Can you explain this a bit?  Why are we temporarily setting the decimal laces value and then setting it back?

Reply:
This is used when the main region is scaled, because otherwise the sub regions are not scaled correctly.

Reply:
But why a decimalplaces of '3'?  And do you know why the other regions aren't scaled when the main region is scaled?  I find that a bit surprising.

Reply:
if the main region is changed, it is only done in whole numbers, the same applies to the subregions.
if the main region is scaled up by 1 and the subregions are 3 times smaller, the result is 0.33, which is rounded down to 0.
if we set the DecimalPlaces temporarily to 2 or higher, the 0.33 is accepted and the size ratio remains.

``` c#
public double X
        {
            get => _x;
            set
            {
                _x = Math.Round(value, DecimalPlaces);
                if (_x < 0)
                    _x = 0;

                if (OwnedTexture != null && OwnedTexture.ImageWidth != 0)
                {
                    if (_x + Width > OwnedTexture.ImageWidth)
                        _x = OwnedTexture.ImageWidth - Width;
                }

                OnPropertyChanged(nameof(X));
            }
        }
```

Reply:
I think this is acceptable for now but we do this in a few places, so it'd be nice if we could encapsulate it somehow.  Additionally, maybe more importantly, it feels like we could still get rounding issues if we needed better precision.  Assuming '3' will work is unfortunate.